scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/TreeOrderedBuf.scala

```diff
@@ -0,0 +1,299 @@
+        _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.DynamicLen(${f.asInstanceOf[FastLengthCalculation[c.type]].t})
+        """)
+        case m: MaybeLengthCalculation[_] => Some(m.asInstanceOf[MaybeLengthCalculation[c.type]].t)
+      }
+
+      fnBodyOpt.map { fnBody =>
+        q"""
+        @inline private[this] def payloadLength($element: $T): _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.MaybeLength = {
+          $fnBody
+        }
+        """
+      }.getOrElse(q"()")
+    }
+
+    def binaryLengthGen(typeName: Tree): (Tree, Tree) = {
+      val tempLen = freshT("tempLen")
+      val lensLen = freshT("lensLen")
+      val element = freshT("element")
+      val callDynamic = (q"""override def staticSize: Option[Int] = None""",
+        q"""
+
+      override def dynamicSize($element: $typeName): Option[Int] = {
+        val $tempLen = payloadLength($element) match {
+          case _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.NoLengthCalculation => None
+          case _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.ConstLen(l) => Some(l)
+          case _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.DynamicLen(l) => Some(l)
+        }
+        (if ($tempLen.isDefined) {
+          // Avoid a closure here while we are geeking out
+          val innerLen = $tempLen.get
+          val $lensLen = posVarIntSize(innerLen)
+          Some(innerLen + $lensLen)
+       } else None): Option[Int]
+     }
+      """)
+      t.length(q"$element") match {
+        case _: NoLengthCalculationAvailable[_] => (q"""
+          override def staticSize: Option[Int] = None""", q"""
+          override def dynamicSize($element: $typeName): Option[Int] = None""")
+        case const: ConstantLengthCalculation[_] => (q"""
+          override val staticSize: Option[Int] = Some(${const.toInt})""", q"""
+          override def dynamicSize($element: $typeName): Option[Int] = staticSize""")
+        case f: FastLengthCalculation[_] => callDynamic
+        case m: MaybeLengthCalculation[_] => callDynamic
+      }
+    }
+
+    def putFnGen(outerbaos: TermName, element: TermName) = {
+      val baos = freshT("baos")
+      val len = freshT("len")
+      val oldPos = freshT("oldPos")
+
+      /**
+       * This is the worst case: we have to serialize in a side buffer
+       * and then see how large it actually is. This happens for cases, like
+       * string, where the cost to see the serialized size is not cheaper than
+       * directly serializing.
+       */
+      val noLenCalc = q"""
+      // Start with pretty big buffers because reallocation will be expensive
+      val $baos = new _root_.java.io.ByteArrayOutputStream(256)
+      ${t.put(baos, element)}
+      val $len = $baos.size
+      $outerbaos.writePosVarInt($len)
+      $baos.writeTo($outerbaos)
```

Comment:
this writes to a side buffer, then copies it into the final buffer right?
Parquet handles this case in a few places by using something similar to a BAOS that instead just collects byte arrays instead of copying them into an internal array. Not sure if that would be better or not, depends on if you most write big chunks or if you mostly write small chunks.